File: application_android/app/src/main/java/com/messenger/app/data/encryption/history/HistoryKeyring.kt
package com.messenger.app.data.encryption.history

/**
 * One (chatId, rootVersion) -> historyRoot binding.
 *
 * A chat may hold several versions simultaneously: after a rotation, older archived messages still
 * need their older root to open, so versions accumulate rather than replace.
 */
class HistoryRootEntry(
    val chatId: String,
    val rootVersion: Int,
    val root: ByteArray,
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is HistoryRootEntry) return false
        return chatId == other.chatId &&
            rootVersion == other.rootVersion &&
            root.contentEquals(other.root)
    }

    override fun hashCode(): Int =
        (chatId.hashCode() * 31 + rootVersion) * 31 + root.contentHashCode()

    /** Never render root material, including accidentally through logging or debuggers. */
    override fun toString(): String = "HistoryRootEntry(chatId=" + chatId +
        ", rootVersion=" + rootVersion + ", root=<redacted 32B>)"
}

/**
 * Minimal versioned keyring for history roots.
 *
 * Deliberately small: Gate 2 defines only the representation. How it is sealed, where it is stored,
 * and when roots rotate are Gate 3 decisions, so nothing here touches the vault, Room, or the
 * network. The representation has no field for a password, an MLS secret, or a device identity
 * private key, and must never gain one - those belong to different trust domains.
 *
 * Wire format (all integers fixed-width big-endian):
 *
 *     u8    formatVersion            (must be 1)
 *     u32be entryCount
 *     repeated entryCount times:
 *       u32be chatIdLen
 *       bytes chatId                 (UTF-8, non-empty)
 *       u32be rootVersion            (>= 1)
 *       u8    rootLen                (must be exactly 32)
 *       bytes root
 *
 * Entries are sorted by (chatId, rootVersion) before encoding, so serialisation is canonical:
 * the same set of entries always produces byte-identical output regardless of insertion order.
 */
class HistoryKeyring private constructor(val entries: List<HistoryRootEntry>) {

    fun encode(): ByteArray {
        val out = java.io.ByteArrayOutputStream()
        out.write(FORMAT_VERSION)
        out.write(HistoryCrypto.u32be(entries.size))
        for (e in entries) {
            val chat = e.chatId.toByteArray(Charsets.UTF_8)
            out.write(HistoryCrypto.u32be(chat.size))
            out.write(chat)
            out.write(HistoryCrypto.u32be(e.rootVersion))
            out.write(e.root.size)
            out.write(e.root)
        }
        return out.toByteArray()
    }

    /** Highest known version for a chat, or null if the chat has no root. */
    fun latest(chatId: String): HistoryRootEntry? =
        entries.filter { it.chatId == chatId }.maxByOrNull { it.rootVersion }

    fun find(chatId: String, rootVersion: Int): HistoryRootEntry? =
        entries.firstOrNull { it.chatId == chatId && it.rootVersion == rootVersion }

    override fun equals(other: Any?): Boolean =
        this === other || (other is HistoryKeyring && entries == other.entries)

    override fun hashCode(): Int = entries.hashCode()

    companion object {
        const val FORMAT_VERSION = 1

        /** Validates and canonicalises. Every rejection below is also a decode-time rejection. */
        fun of(entries: List<HistoryRootEntry>): Result<HistoryKeyring> {
            val seen = HashSet<Pair<String, Int>>()
            for (e in entries) {
                if (e.chatId.isEmpty()) {
                    return fail("chatId must not be empty")
                }
                if (e.rootVersion < 1) {
                    return fail("rootVersion must be >= 1, was " + e.rootVersion)
                }
                if (e.root.size != HistoryCrypto.ROOT_BYTES) {
                    return fail(
                        "history root must be exactly " + HistoryCrypto.ROOT_BYTES +
                            " bytes, was " + e.root.size
                    )
                }
                if (!seen.add(e.chatId to e.rootVersion)) {
                    return fail(
                        "duplicate (chatId, rootVersion) entry for version " + e.rootVersion
                    )
                }
            }
            val canonical = entries.sortedWith(
                compareBy({ it.chatId }, { it.rootVersion })
            )
            return Result.success(HistoryKeyring(canonical))
        }

        fun decode(bytes: ByteArray): Result<HistoryKeyring> {
            val r = Reader(bytes)
            try {
                val version = r.u8()
                if (version != FORMAT_VERSION) {
                    return fail("unsupported keyring format version " + version)
                }
                val count = r.u32()
                if (count < 0) return fail("negative entry count")
                // Each entry needs at least 4 + 1 + 4 + 1 + 32 bytes; reject absurd counts before
                // allocating, so a corrupt length cannot drive a huge allocation.
                if (count.toLong() * 9L > bytes.size.toLong()) {
                    return fail("entry count " + count + " exceeds available bytes")
                }
                val parsed = ArrayList<HistoryRootEntry>(count)
                repeat(count) {
                    val chatId = String(r.bytes(r.u32()), Charsets.UTF_8)
                    val rootVersion = r.u32()
                    val rootLen = r.u8()
                    val root = r.bytes(rootLen)
                    parsed.add(HistoryRootEntry(chatId, rootVersion, root))
                }
                if (r.remaining() != 0) {
                    return fail("trailing bytes after keyring: " + r.remaining())
                }
                return of(parsed)
            } catch (e: IllegalArgumentException) {
                return Result.failure(e)
            }
        }

        private fun fail(message: String): Result<HistoryKeyring> =
            Result.failure(IllegalArgumentException(message))
    }

    /** Strict big-endian reader; every overrun is an explicit failure, never a silent clamp. */
    private class Reader(private val buf: ByteArray) {
        private var pos = 0

        fun remaining(): Int = buf.size - pos

        fun u8(): Int {
            need(1)
            return buf[pos++].toInt() and 0xFF
        }

        fun u32(): Int {
            need(4)
            val v = ((buf[pos].toInt() and 0xFF) shl 24) or
                ((buf[pos + 1].toInt() and 0xFF) shl 16) or
                ((buf[pos + 2].toInt() and 0xFF) shl 8) or
                (buf[pos + 3].toInt() and 0xFF)
            pos += 4
            return v
        }

        fun bytes(n: Int): ByteArray {
            if (n < 0) throw IllegalArgumentException("negative length " + n)
            need(n)
            val out = buf.copyOfRange(pos, pos + n)
            pos += n
            return out
        }

        private fun need(n: Int) {
            if (remaining() < n) {
                throw IllegalArgumentException(
                    "truncated keyring: need " + n + " bytes, have " + remaining()
                )
            }
        }
    }
}

File: windows_app/src/crypto/historykeyring.cpp
#include "historykeyring.h"
#include "historycrypto.h"

#include <algorithm>

namespace {

void setError(QString* error, const QString& message) {
    if (error) *error = message;
}

/**
 * Strict big-endian reader. Every overrun is an explicit failure rather than a silent clamp, which
 * is what stops a corrupt length from being read as a shorter valid keyring.
 */
class Reader {
public:
    explicit Reader(const QByteArray& buf) : m_buf(buf) {}

    int remaining() const { return m_buf.size() - m_pos; }
    bool failed() const { return m_failed; }

    int u8() {
        if (!need(1)) return 0;
        return static_cast<int>(static_cast<unsigned char>(m_buf.at(m_pos++)));
    }

    int u32() {
        if (!need(4)) return 0;
        const auto b = [&](int i) {
            return static_cast<unsigned int>(static_cast<unsigned char>(m_buf.at(m_pos + i)));
        };
        const unsigned int v = (b(0) << 24) | (b(1) << 16) | (b(2) << 8) | b(3);
        m_pos += 4;
        // Reinterpreted as signed, exactly as Android's Int-returning reader does, so an oversized
        // field arrives here as a negative number and is rejected by the same checks.
        return static_cast<int>(v);
    }

    QByteArray bytes(int n) {
        if (n < 0) { m_failed = true; return {}; }
        if (!need(n)) return {};
        const QByteArray out = m_buf.mid(m_pos, n);
        m_pos += n;
        return out;
    }

private:
    bool need(int n) {
        if (m_failed) return false;
        if (remaining() < n) { m_failed = true; return false; }
        return true;
    }

    const QByteArray& m_buf;
    int m_pos = 0;
    bool m_failed = false;
};

} // namespace

bool HistoryKeyring::of(const QVector<HistoryRootEntry>& entries,
                        HistoryKeyring& out,
                        QString* error) {
    QVector<QPair<QString, int>> seen;
    seen.reserve(entries.size());
    for (const HistoryRootEntry& e : entries) {
        if (e.chatId.isEmpty()) {
            setError(error, QStringLiteral("chatId must not be empty"));
            return false;
        }
        if (e.rootVersion < 1) {
            setError(error, QStringLiteral("rootVersion must be >= 1, was %1").arg(e.rootVersion));
            return false;
        }
        if (e.root.size() != HistoryCrypto::ROOT_BYTES) {
            setError(error, QStringLiteral("history root must be exactly %1 bytes, was %2")
                                .arg(HistoryCrypto::ROOT_BYTES)
                                .arg(e.root.size()));
            return false;
        }
        const QPair<QString, int> key(e.chatId, e.rootVersion);
        if (seen.contains(key)) {
            setError(error, QStringLiteral("duplicate (chatId, rootVersion) entry for version %1")
                                .arg(e.rootVersion));
            return false;
        }
        seen.append(key);
    }

    QVector<HistoryRootEntry> canonical = entries;
    std::sort(canonical.begin(), canonical.end(),
              [](const HistoryRootEntry& a, const HistoryRootEntry& b) {
                  if (a.chatId != b.chatId) return a.chatId < b.chatId;
                  return a.rootVersion < b.rootVersion;
              });
    out = HistoryKeyring(std::move(canonical));
    return true;
}

QByteArray HistoryKeyring::encode() const {
    QByteArray out;
    out.append(static_cast<char>(FORMAT_VERSION));
    out.append(HistoryCrypto::u32be(static_cast<quint32>(m_entries.size())));
    for (const HistoryRootEntry& e : m_entries) {
        const QByteArray chat = e.chatId.toUtf8();
        out.append(HistoryCrypto::u32be(static_cast<quint32>(chat.size())));
        out.append(chat);
        out.append(HistoryCrypto::u32be(static_cast<quint32>(e.rootVersion)));
        out.append(static_cast<char>(e.root.size()));
        out.append(e.root);
    }
    return out;
}

bool HistoryKeyring::decode(const QByteArray& bytes, HistoryKeyring& out, QString* error) {
    Reader r(bytes);

    const int version = r.u8();
    if (r.failed()) {
        setError(error, QStringLiteral("truncated keyring"));
        return false;
    }
    if (version != FORMAT_VERSION) {
        setError(error, QStringLiteral("unsupported keyring format version %1").arg(version));
        return false;
    }

    const int count = r.u32();
    if (r.failed()) {
        setError(error, QStringLiteral("truncated keyring"));
        return false;
    }
    if (count < 0) {
        setError(error, QStringLiteral("negative entry count"));
        return false;
    }
    // Each entry needs at least 4 + 1 + 4 + 1 + 32 bytes; reject absurd counts before allocating so
    // a corrupt length cannot drive a huge allocation.
    if (static_cast<qint64>(count) * 9LL > static_cast<qint64>(bytes.size())) {
        setError(error, QStringLiteral("entry count %1 exceeds available bytes").arg(count));
        return false;
    }

    QVector<HistoryRootEntry> parsed;
    parsed.reserve(count);
    for (int i = 0; i < count; ++i) {
        HistoryRootEntry e;
        const QByteArray chat = r.bytes(r.u32());
        e.chatId = QString::fromUtf8(chat);
        e.rootVersion = r.u32();
        const int rootLen = r.u8();
        e.root = r.bytes(rootLen);
        if (r.failed()) {
            setError(error, QStringLiteral("truncated keyring"));
            return false;
        }
        parsed.append(e);
    }

    if (r.remaining() != 0) {
        setError(error, QStringLiteral("trailing bytes after keyring: %1").arg(r.remaining()));
        return false;
    }
    return of(parsed, out, error);
}

bool HistoryKeyring::latest(const QString& chatId, HistoryRootEntry& out) const {
    bool found = false;
    for (const HistoryRootEntry& e : m_entries) {
        if (e.chatId != chatId) continue;
        if (!found || e.rootVersion > out.rootVersion) {
            out = e;
            found = true;
        }
    }
    return found;
}

bool HistoryKeyring::find(const QString& chatId, int rootVersion, HistoryRootEntry& out) const {
    for (const HistoryRootEntry& e : m_entries) {
        if (e.chatId == chatId && e.rootVersion == rootVersion) {
            out = e;
            return true;
        }
    }
    return false;
}

int HistoryKeyring::latestVersion(const QString& chatId) const {
    HistoryRootEntry e;
    return latest(chatId, e) ? e.rootVersion : 0;
}

bool HistoryKeyring::withAddedVersion(const QString& chatId,
                                      int rootVersion,
                                      const QByteArray& root,
                                      HistoryKeyring& out,
                                      QString* error) const {
    QVector<HistoryRootEntry> next = m_entries;
    HistoryRootEntry add;
    add.chatId = chatId;
    add.rootVersion = rootVersion;
    add.root = root;
    next.append(add);
    return of(next, out, error);
}
